packages/preview-middleware-client/src/cpe/changes/generic-change.ts

```diff
@@ -0,0 +1,469 @@
+    | ConfigChange
+    | V2ConfigChange;
+
+interface BaseChange extends ChangeDefinition {
+    creation: string;
+}
+
+interface ChangeContent {
+    property: string;
+    newValue: string;
+    newBinding: string;
+}
+
+export interface AddXml extends BaseChange {
+    changeType: typeof ADD_XML_CHANGE;
+    content: {
+        targetAggregation: string;
+        fragmentPath: string;
+    };
+}
+export interface RenameChange extends BaseChange {
+    changeType: typeof RENAME_CHANGE;
+    texts: {
+        newText: {
+            value: string;
+            type: string;
+        };
+    };
+}
+export interface ConfigChange extends BaseChange {
+    changeType: typeof MANIFEST_V4_CHANGE;
+    propertyName: string;
+    content: ConfigurationChangeContent;
+}
+
+export interface V2ConfigChange extends BaseChange {
+    changeType: typeof MANIFEST_V4_CHANGE;
+    propertyName: string;
+    content: {
+        entityPropertyChange: {
+            propertyPath: string;
+            propertyValue: Record<string, string>;
+        };
+        parentPage: {
+            component: string;
+            entitySet: string;
+        };
+    };
+}
+
+interface ConfigurationChangeContent {
+    page: string;
+    entityPropertyChange: {
+        propertyPath: string;
+        operation: 'UPSERT' | 'DELETE' | 'INSERT' | 'UPDATE';
+        propertyValue: string | Record<string, string>;
+    };
+}
+
+export interface NewAnnotationFileChange extends BaseChange {
+    changeType: typeof ADD_NEW_ANNOTATION_FILE_CHANGE;
+    content: {
+        dataSourceId: string;
+        dataSource: {
+            [key: string]: {
+                uri: string;
+            };
+        };
+    };
+}
+
+export interface MoveControlsChange extends BaseChange {
+    changeType: typeof MOVE_CHANGE;
+    content: {
+        movedElements: {
+            selector: {
+                id: string;
+            };
+            sourceIndex: string;
+            targetIndex: string;
+        }[];
+        target: {
+            selector: {
+                id: string;
+            };
+        };
+    };
+}
+
+export interface PropertyChange extends BaseChange {
+    changeType: typeof PROPERTY_CHANGE | typeof PROPERTY_BINDING_CHANGE;
+    controlId: string;
+    controlName: string;
+    propertyName: string;
+    content: ChangeContent;
+}
+
+interface ChangeHandlerOptions {
+    appComponent: AppComponent;
+    textBundle: TextBundle;
+    configPropertyControlIdMap?: Map<string, string[]>;
+}
+
+interface GenericChangeHandlerReturnType {
+    changeTitle: string;
+    controlId?: string | string[];
+    changeType?: string;
+    configPath?: string;
+    properties: { label: string; value: PropertyValue | Record<string, unknown>; displayValueWithIcon?: boolean }[];
+}
+
+interface GenericChangeHandler {
+    [key: string]: (
+        change: GenericChange,
+        options: ChangeHandlerOptions
+    ) => Promise<GenericChangeHandlerReturnType> | GenericChangeHandlerReturnType;
+}
```

Comment:
We can narrow down the changes within the handlers by restricting that `changeType` must match the key in the map. Then we don't need the unsafe assertions in every handler.
```suggestion
export type ChangeHandler<Change> = (
    change: Change,
    options: ChangeHandlerOptions
) => Promise<GenericChangeHandlerReturnType> | GenericChangeHandlerReturnType
type GenericChangeHandler = {
    [Change in GenericChange as Change['changeType']]: ChangeHandler<Change>;
};```

Reply:
changed